### PR TITLE
RR-760: Incorporate TypeScript template changes

### PR DESCRIPTION
**Pull request by leejacobson-moj**
## Overview

This PR incorporates the latest changes from the TypeScript project, following the recommendation here: https://mojdt.slack.com/archives/C69NWE339/p1694009011413449

We should do this regularly to prevent our project becoming too misaligned from the template.

Notable changes in this PR:
- Removes deprecated `/auth/api/user/me` HMPPS Auth endpoint in favour of Manage Users API, `/users/me`
- Adds support for HMPPS Audit
- Adds support for automatic retries of failed `GET` & `DELETE` requests
- Adds new service `/info` endpoint
- Updates `/health` endpoint
- Improved appInsights reporting
- Dependency bumps

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,8 +1,9 @@
 # Stage: base image
-FROM node:20.10-bullseye-slim as base
+FROM node:20.11-bookworm-slim as base
 
-ARG BUILD_NUMBER=1_0_0
-ARG GIT_REF=not-available
+ARG BUILD_NUMBER
+ARG GIT_REF
+ARG GIT_BRANCH
```
Comment (leejacobson-moj) [inline Dockerfile:6]: CircleCI passes these into the docker build:

Screenshot at Mar 28 13-07-47

File: README.md [hunk 2]
```diff
@@ -80,6 +80,7 @@ Note: Using `nvm` (or [fnm](https://github.com/Schniz/fnm)), run `nvm install --
 Create a `.env` which should override environment variables required to run locally:
 ```properties
 HMPPS_AUTH_URL=http://localhost:9090/auth
+MANAGE_USERS_API_URL=http://localhost:9091/manage-users-api
```
Comment (leejacobson-moj) [inline README.md:83]: HMPPS Auth endpoint, `/auth/api/user/me` is deprecated. This info now comes from the manage users API, (`/users/me`)

File: docker-compose.yml [hunk 3]
```diff
@@ -32,6 +32,7 @@ services:
     ports:
       - "3000:3000"
     environment:
+      - PRODUCT_ID=DPS027
```
Comment (leejacobson-moj) [inline docker-compose.yml:35]: Used by the `/info` endpoint.

File: server/data/cache/tokenStore.ts
Comment (leejacobson-moj) [inline server/data/cache/tokenStore.ts:1]: Moved to `server/data/cache/tokenStore.ts` to align with template

File: server/data/prisonRegisterClient.test.ts [hunk 4]
```diff
@@ -58,7 +58,7 @@ describe('prisonRegisterClient', () => {
         userMessage: 'An unexpected error occurred',
         developerMessage: 'An unexpected error occurred',
       }
-      prisonRegisterApi.get('/prisons').reply(500, expectedResponseBody)
+      prisonRegisterApi.get('/prisons').thrice().reply(500, expectedResponseBody)
```
Comment (leejacobson-moj) [inline server/data/prisonRegisterClient.test.ts:61]: On error `RestClient` will now retry `GET` requests twice for a total call count of three. All three API calls need to return a 500 error for this test to pass.

File: server/data/redisClient.ts
Comment (leejacobson-moj) [inline server/data/redisClient.ts:1]: Moved to align with template

File: server/views/pages/prisonerList/partials/searchTable.njk [hunk 5]
```diff
@@ -14,7 +14,7 @@
       <input type="hidden" name="statusFilter" value="{{ statusFilter }}" />
       <input type="hidden" name="page" value="{{ page }}" />
 
-      <table class="govuk-table" data-module="moj-sortable-table" data-qa="prisoner-list-results-table">
```
Comment (leejacobson-moj) [inline server/views/pages/prisonerList/partials/searchTable.njk:17]: MoJ Frontend JS now initialises on all pages and we disable it for this table.

File: server/data/hmppsAuditClient.ts
Comment (leejacobson-moj) [inline server/data/hmppsAuditClient.ts:1]: New Audit client.

File: server/middleware/setUpHealthChecks.ts [hunk 6]
```diff
@@ -21,5 +21,19 @@ export default function setUpHealthChecks(applicationInfo: ApplicationInfo): Rou
     }),
   )
 
+  router.get('/info', (req, res) => {
```
Comment (leejacobson-moj) [inline server/middleware/setUpHealthChecks.ts:24]: Added support for new `/info` endpoint.

File: server/data/educationAndWorkPlanClient.test.ts [hunk 7]
```diff
@@ -101,7 +101,7 @@ describe('educationAndWorkPlanClient', () => {
         userMessage: 'An unexpected error occurred',
         developerMessage: 'An unexpected error occurred',
       }
-      educationAndWorkPlanApi.get(`/action-plans/${prisonNumber}`).reply(500, expectedResponseBody)
+      educationAndWorkPlanApi.get(`/action-plans/${prisonNumber}`).thrice().reply(500, expectedResponseBody)
```
Comment (nathanrussell-moj-digital) [inline server/data/educationAndWorkPlanClient.test.ts:104]: Ah ha, to account for the automatic retries ? 👍

Comment (leejacobson-moj) [reply server/data/educationAndWorkPlanClient.test.ts:104]: Yep, exactly.

**Pull request closed without merging**